app/bulk_resource/views.py

```diff
@@ -92,10 +93,95 @@ def upload_data():
+            row.csv_body_cells[csv_container.address_column_index].data
+        g = geocoder.google(address)
+        resource = Resource.query.filter_by(name=name).first()
+
+        if resource is None:
+            resource = Resource(
+                name=name,
+                address=address,
+                latitude=g.latlng[0],
+                longitude=g.latlng[1]
+            )
+            db.session.add(resource)
+
+        else:
+            resource.latitude = g.latlng[0]
+            resource.longitude = g.latlng[1]
+
+        for i, cell in enumerate(row.csv_body_cells):
+            if i not in csv_container.required_column_indices():
+                descriptor_name = \
+                    csv_container.csv_header_row.csv_header_cells[i].data
+                descriptor = Descriptor.query.filter_by(
+                    name=descriptor_name
+                ).first()
+                if descriptor is None:
+                    errors.append('Descriptor {} is invalid'.format(descriptor_name))
+                    continue
+
+                text_association = [td for td in resource.text_descriptors if td.descriptor_id == descriptor.id]
+                text_association = text_association[0] if text_association else None
+                option_association = [od for od in resource.option_descriptors if od.descriptor_id == descriptor.id]
+                option_association = option_association[0] if option_association else None
+
+                if text_association is not None:
+                    db.session.delete(text_association)
+                elif option_association is not None:
+                    db.session.delete(option_association)
+
+                values = list(descriptor.values)
+                assocValues = []
+                if len(descriptor.values) == 0:  # text descriptor
+                    association_class = TextAssociation
+                    assocValues.append(cell.data)
+                    keyword = 'text'
+                else:  # option descriptor
+                    association_class = OptionAssociation
+                    for s in cell.data.split(';'):
+                        if s in values:
+                            assocValues.append(values.index(s))
+                    keyword = 'option'
+                for value in assocValues:
+                    arguments = {
+                        'resource_id': resource.id,
+                        'descriptor_id': descriptor.id,
+                        keyword: value,
+                        'resource': resource,
+                        'descriptor': descriptor
+                    }
+                    new_association = association_class(**arguments)
+                    db.session.add(new_association)
+
+    if len(errors) > 0:
+        db.session.rollback()
+        for error in errors:
+            flash(error, 'error')
+        abort(400)
+
+    db.session.commit()
+    return redirect(url_for('single_resource.index'))
```

Comment:
i'm thinking that if we add new descriptors in this csv upload we would also want to go to same sort of workflow as we do for reset in terms of choosing the type of descriptor it is?